src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -47,16 +47,29 @@ public void BeginOperation(CancellationToken cancellationToken, Action<object?>
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            _awaitableState |= AwaitableState.Running;
-
             // Don't register if already completed, we would immediately unregistered in ObserveCancellation
             if (cancellationToken.CanBeCanceled && !IsCompleted)
             {
+                var previousAwaitableState = _awaitableState;
+
+                _cancellationTokenRegistration = cancellationToken.UnsafeRegister(callback, state);
```

Comment:
What do these callbacks do?  Seems like this method is always called while holding a lock, which means this callback could be invoked while holding the lock.  Is that ok?

Reply:
Yes, it's ok. The callback takes the lock (which is already held) and noops.